### ISSUE TITLE
Troubleshoot latency in Cloud Bigtable C++ Client

### ISSUE THREAD

**Issue by coryan** [GoogleCloudPlatform/google-cloud-cpp#192]
Most of the time a client talking to the production server experiences very high latencies, in the order of 80ms [*].  The objective of this bug is to isolate the problem, if necessary file bugs with other components, or determine what needs fixing in the Cloud Bigtable C++ Client.

[*]: In this bug we will try to consistently use ms == milliseconds, and &mu;s == microseconds.

**Comment by coryan** [comment]
## Setup

I created a GCE virtual machine in the same zone (us-east1-c) as a Cloud Bigtable instance.  Then compiled the code as usual:

```console
cd $HOME
git clone ... # e.g. -b quick-benchmark git@github.com:coryan/google-cloud-cpp.git
cd google-cloud-cpp
git submodule update --init --recursive
mkdir .build && cd .build
cmake -DCMAKE_BUILD_TYPE=Release ..
make # -j $(nproc) to parallelize, if your system lacks nproc(1) use the number of cores
```
These benchmarks will connect to the production environment, you need to tell
gRPC where to find the root certificates:

```console
export GRPC_DEFAULT_SSL_ROOTS_FILE_PATH=$HOME/google-cloud-cpp/third_party/grpc/etc/roots.pem
```

These benchmarks use the golang tools for Cloud Bigtable, and a small golang program
to demonstrate that the problems are not in the server or the network:

```console
export GOPATH=$HOME/google-cloud-cpp/bigtable/benchmarks
go get -u cloud.google.com/go/bigtable/cmd/emulator cloud.google.com/go/bigtable/cmd/cbt
go get -u google.golang.org/api/option
go get -u google.golang.org/api/transport
go get -u golang.org/x/net/context
go get -u google.golang.org/grpc
go get -u google.golang.org/genproto/googleapis/bigtable/v2
go build ping_client_go
```

To run the benchmarks you need a Cloud Bigtable instance.  Follow the 
[quickstart guide](https://cloud.google.com/bigtable/docs/quickstart-cbt)
to set this up.  Create the Cloud Bigtable instance in the same zone as your
GCE instances.  Store your project id and instance id in convenience variables:

```console
PROJECT=... # put your project name here (e.g. my-project)
INSTANCE=... # put the instance id here (e.g. test-instance-0)
$GOPATH/bin/cbt -project $PROJECT -instance $INSTANCE createtable table-0 families=cf:maxversions=1
```

**Comment by coryan** [comment]
**Observation:** The Cloud Bigtable C++ client experiences high latency (~80ms)
when contacting the production service.

**Why do we think this:** because the `apply_read_latency_benchmark` reports
high numbers:

```console
$ ./bigtable/apply_read_latency_benchmark $PROJECT $INSTANCE perf-0001 1 60 10000
Populating table ........................................ DONE
name,nsamples,min,p50,p90,p95,p99,p99.9,max
Apply()/Success,224,80849,82583,86576,88167,92221,212068,212068
No results for Apply()/Failure
name,nsamples,min,p50,p90,p95,p99,p99.9,max
ReadRow()/Success,495,80184,82303,85752,87841,92275,105824,105824
No results for ReadRow()/Failure
```

**Comment by coryan** [comment]
Our initial guess was that the code in the Cloud Bigtable C++ client had some
problem.  But we wrote a simple gRPC C++ client and got:

**Observation:** A simple gRPC/C++ client has similar latencies when making
calls to `bigtable.googleapis.com`.

```console
./bigtable/ping_client_cpp $PROJECT $INSTANCE table-0 100
################ request 0 2352699us
################ request 1 94517us
################ request 2 86483us
################ request 3 86441us
[snip ... snip]
################ request 97 86756us
################ request 98 86459us
################ request 99 85986us
Average latency = 109653us
```

There are a number of possible explanations:

1. The Network is slow.
1. Cloud Bigtable is slow.
1. The gRPC C++ code is slow.
1. gRPC/C++ interacts poorly with `bigtable.googleapis.com`

We will eliminate them in turn, from easiest to hardest.

**Comment by coryan** [comment]
## It is not "The Network is slow".

While the benchmark is running on a GCE instance in the same zone as
Cloud Bigtable and `ping(1)` reports good latency:

```console
ping bigtable.googleapis.com
PING googleapis.l.google.com (74.125.141.95) 56(84) bytes of data.
64 bytes from vl-in-f95.1e100.net (74.125.141.95): icmp_seq=1 ttl=52 time=0.678 ms
64 bytes from vl-in-f95.1e100.net (74.125.141.95): icmp_seq=2 ttl=52 time=0.294 ms
64 bytes from vl-in-f95.1e100.net (74.125.141.95): icmp_seq=3 ttl=52 time=0.297 ms
64 bytes from vl-in-f95.1e100.net (74.125.141.95): icmp_seq=4 ttl=52 time=0.269 ms
```

It is still possible that the network latency is higher for the application.
However, a simple gRPC/Go has much better latency:

```console
./ping_client_go $PROJECT $INSTANCE table-0 500
Average latency =  3.240266ms
```

**Comment by coryan** [comment]
### It is not "Cloud Bigtable is slow"

The Go client already eliminated this hypothesis.

**Comment by coryan** [comment]
### It is not "The gRPC C++ code is slow"

This is more difficult to disprove directly, but we can provide several
arguments that eliminate this hypothesis.

**Comment by coryan** [comment]
#### The CPU utilization is low for the gRPC/C++ client

This indicates that the latency is introduced by I/O calls, or the network,
but not by code taking a lot of time to process "something":

```console
$ /usr/bin/time ./bigtable/ping_client_cpp $PROJECT $INSTANCE table-0 100
################ request 0 1449521us
################ request 1 89617us
################ request 2 81598us
################ request 3 81244us
[snip ... snip]
################ request 97 81246us
################ request 98 80986us
################ request 99 81230us
Average latency = 94957us
0.03user 0.02system 0:09.50elapsed 0%CPU (0avgtext+0avgdata 10000maxresident)k
0inputs+0outputs (0major+853minor)pagefaults 0swaps
```

**Comment by coryan** [comment]
#### The latency is low running against the Cloud Bigtable Emulator

The same program reports much lower latency (~100&mu;s) when contacting the
Cloud Bigtable emulator in the localhost.

Running this on a separate shell:

```console
$GOPATH/bin/emulator
```

Setup the test table in the emulator:

```console
BIGTABLE_EMULATOR_HOST=localhost:9000 ../bigtable/benchmarks/bin/cbt -project $PROJECT -instance $INSTANCE createtable table-0 families=cf:maxversions=1
```

And then run the test:

```console
./bigtable/ping_client_cpp $PROJECT $INSTANCE table-0 100 0 localhost:9000
################ request 0 1022us
################ request 1 224us
################ request 2 243us
[snip snip snip]
################ request 97 153us
################ request 98 117us
################ request 99 465us
Average latency = 175us
```

If the code was slow it would be slow for both, though admittedly the code paths
are slightly different, as the local version does not involve encryption or
authentication.

**Comment by coryan** [comment]
#### strace(1) results indicate that the delay may be in the underlying messages

We can use `strace` to see that after sending the request with
`sendmsg()` the corresponding `recvmsg()` takes many milliseconds. This time
we configure the test to pause for 2,000 milliseconds after each RPC:
 
```console
strace -r -e trace=network,epoll_pwait ./bigtable/ping_client_cpp $PROJECT $INSTANCE table-0 50 2000 >ping_cpp.log 2>&1
```

And look at one of the requests to see the delays:

```console
 grep -B 10 'request 17 ' ping_cpp.log 
################ request 16 81467us
     2.000741 sendmsg(5, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\27\3\3\1_\0\0\0\0\0\0\0$\260&\357\234 \315\v\234a\210\277Ye|\261\374\227\31\25"..., iov_len=356}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 356 <0.000073>
     0.000177 epoll_pwait(3, [{EPOLLOUT, {u32=4160755344, u64=140415231596176}}], 100, -1, NULL, 8) = 1 <0.000029>
     0.000097 epoll_pwait(3, [{EPOLLIN|EPOLLOUT, {u32=4160755344, u64=140415231596176}}], 100, -1, NULL, 8) = 1 <0.000355>
     0.000426 recvmsg(5, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\27\3\3\0)\0\0\0\0\0\0\0F\370\264\267\27\332\362\211b\200h\201\216U\325)\f?X\241"..., iov_len=7401}, {iov_base="", iov_len=1280}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, 0) = 46 <0.000037>
     0.000096 recvmsg(5, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable) <0.000020>
     0.000059 epoll_pwait(3, [{EPOLLIN|EPOLLOUT, {u32=4160755344, u64=140415231596176}}], 100, -1, NULL, 8) = 1 <0.080302>
     0.080425 recvmsg(5, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\27\3\3\0@\0\0\0\0\0\0\0G\312\372\211\263\7\270\351\33\25\214p\326W\\\304\260\203\261\26"..., iov_len=1280}, {iov_base="", iov_len=7355}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, 0) = 169 <0.000043>
     0.000164 recvmsg(5, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable) <0.000056>
     0.000122 sendmsg(5, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\27\3\3\0)\0\0\0\0\0\0\0%AR\234\351S\272\210k\327\352j\352\325Hdgq\324\251"..., iov_len=46}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 46 <0.000039>
################ request 17 81780us
```

Notice how the application (really, the gRPC library) blocks for ~80ms in the
third `epoll_pwait()` and then calls
`recvmsg()`.  This indicates that the application was waiting for the message
from the server, and not performing work locally.

In addition, note how the second `epoll_pwait()` does not block.  This indicates
that a response was already available in the socket, probably to the trailing
`sendmsg()` request which is made just before the RPC completes.

It is also interesting that the gRPC protocol seems to require two messages
per RPC: there are two successful sendmsg() calls, and two successful
`recvmsg()` calls in each RPC:

```console
fgrep -c 'sendmsg(5,' ping_cpp.log 
104

fgrep  'recvmsg(5,' ping_cpp.log | grep -c -v EAGAIN
92
```

**Comment by coryan** [comment]
These files are examples of a client performing well, though only for a few iterations, it happens rarely, and I have not been able to isolate the reason.

[fast.zip](https://github.com/GoogleCloudPlatform/google-cloud-cpp/files/1650571/fast.zip)

**Comment by coryan** [comment]
A typical tcpdump and strace capture for the program running.

[slow.zip](https://github.com/GoogleCloudPlatform/google-cloud-cpp/files/1650576/slow.zip)

**Comment by coryan** [comment]
The code to demonstrate these problems is not merged yet, it can be found in [my fork](https://github.com/coryan/google-cloud-cpp/tree/quick-benchmark) until we merge it.

**Comment by coryan** [comment]
@mbrukman please see above.  I am pretty sure the latency problems are *not* in our code, and if they are in the gRPC/C++ code it is something that only manifests itself when talking to `googleapis.com`.  Do you have a good contact to send this to?

**Comment by coryan** [comment]
A (mostly) statically linked binary for Ubuntu:

[ping_client_cpp.gz](https://github.com/GoogleCloudPlatform/google-cloud-cpp/files/1650621/ping_client_cpp.gz)

For those that want to run the tests without having to compile all the code (you still need to setup the Cloud Bigtable instance though).

**Comment by mbrukman** [comment]
@coryan — https://github.com/grpc/grpc is the repo for the C-based implementation of gRPC, which is what we're using. You can file an issue in that repo and refer them to this issue for more details, if you would like to go that route.

Alternatively, we can find the right contacts and email them directly, if you prefer that instead. My guess is that @sduskis, @igorbernstein2, and @garye may know whom to talk to on the gRPC team about this question and can introduce you to the right set of folks.

**Comment by coryan** [comment]
An additional note on how to capture packets and program traces, just so I do not forget:

```console
$ sudo tcpdump -w ping_cpp.remote.grpc-trace.pcap dst bigtable.googleapis.com or src bigtable.googleapis.com &
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
$ strace -EGRPC_VERBOSITY=DEBUG -EGRPC_TRACE=all,-timer,-api -ttt -T -e trace=network,epoll_pwait ./bigtable/ping_client_cpp $PROJECT $INSTANCE table-0 50 2000 >ping_cpp.remote.grpc-trace.log 2>&1
```

Then bring `tcpdump` to the foreground and kill it with `^C` (or whatever your INTR key is).  To see the tcpdump use:

```console
$ tcpdump -n -r ping_cpp.remote.grcp-trace.pcap
```

**Comment by coryan** [comment]
This was resolved by #198.

**Comment by randxie** [comment]
@coryan Thanks for the detailed investigation. Recently, I also noticed there's performance degradation in the cloud bigtable c++ client. By any chance, have you or the grpc team figured out the root cause?
